crates/bitwarden-crypto/src/store/backend/implementation/linux_memfd_secret.rs

```diff
@@ -0,0 +1,111 @@
+use std::{mem::MaybeUninit, ptr::NonNull, sync::OnceLock};
+
+use super::{
+    slice_backend::{SliceBackend, SliceLike},
+    KeyRef,
+};
+
+// This is an in-memory key store that is protected by memfd_secret on Linux 5.14+.
+// This should be secure against memory dumps from anything except a malicious kernel driver.
+// Note that not all 5.14+ systems have support for memfd_secret enabled, so
+// LinuxMemfdSecretKeyStore::new returns an Option.
+pub(crate) type LinuxMemfdSecretBackend<Key> = SliceBackend<Key, MemfdSecretImplKeyData>;
+
+pub(crate) struct MemfdSecretImplKeyData {
+    ptr: std::ptr::NonNull<[u8]>,
+    capacity: usize,
+}
+
+// For Send+Sync to be safe, we need to ensure that the memory is only accessed mutably from one
+// thread. To do this, we have to make sure that any funcion in `MemfdSecretImplKeyData` that
+// accesses the pointer mutably is defined as &mut self, and that the pointer is never copied or
+// moved outside the struct.
+unsafe impl Send for MemfdSecretImplKeyData {}
+unsafe impl Sync for MemfdSecretImplKeyData {}
+
+impl Drop for MemfdSecretImplKeyData {
+    fn drop(&mut self) {
+        unsafe {
+            memsec::free_memfd_secret(self.ptr);
+        }
+    }
+}
+
+impl<Key: KeyRef> SliceLike<Key> for MemfdSecretImplKeyData {
+    fn is_available() -> bool {
+        static IS_SUPPORTED: OnceLock<bool> = OnceLock::new();
+
+        *IS_SUPPORTED.get_or_init(|| unsafe {
+            let Some(ptr) = memsec::memfd_secret_sized(1) else {
+                return false;
+            };
+            memsec::free_memfd_secret(ptr);
+            true
+        })
+    }
+
+    fn with_capacity(capacity: usize) -> Self {
+        let entry_size = std::mem::size_of::<Option<(Key, Key::KeyValue)>>();
+
+        unsafe {
+            let ptr: NonNull<[u8]> = memsec::memfd_secret_sized(capacity * entry_size)
+                .expect("memfd_secret_sized failed");
+
+            // Initialize the array with Nones using MaybeUninit
+            let uninit_slice: &mut [MaybeUninit<_>] = std::slice::from_raw_parts_mut(
+                ptr.as_ptr() as *mut MaybeUninit<Option<(Key, Key::KeyValue)>>,
+                capacity,
+            );
+            for elem in uninit_slice {
+                elem.write(None);
+            }
```

Comment:
This is probably unsafe? https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html